Display the new policies on the site instead of linking to github

levitte opened the pull request:
This adds the necessary files and procedures to generate HTML from policis

The checked out directories $(CHECKOUTS)/technical-policies/policies
and $(CHECKOUTS)/general-policies/policies contain files in markdown
format, which get converted to HTML using pandoc with the new template,
and appropriate dirdata.yaml.

The base infrastructure in policies/general and policies/technical is also
created, and the majority of the links in policies/sidebar.shtml are moved
to policies/general/sidebar.shtml and policies/general/dirdata.yaml.

Several old policy files are dropped in favor of those generated in those
generated in policies/general/ and policies/technical/.  Some older policies
still remain, pending replacements.

Fixes #306

levitte committed:
Rework the markdown to HTML script and template

The pandoc template inherits all the HTML that was included with
server side includes.  For things that depend on the directory, we use
metadata variables, which are provided externally, or as part of the
markdown origin.

bin/md-to-html5 is changed accordingly.  Among other things, it checks
for dirdata.yaml in the directory of the markdown origin, and includes
it into the build as well.

dirdata.yaml is expected to contain metadata that applies to all
generated HTML files in the containing directory.  Two metadata
attributes are expected:

- `breadcrumbs`, to give an indication of where the user is, for every
  page.  If the HTML to be produced isn't the index file, it will
  automatically be added to the breadcrumbs by the pandoc template.
- `sidebar`, which will be the content of the sidebar

Both of these attributes take strings that are interpreted as markdown

levitte commented:
I've had ~2018bda3665b5874316e6deac35e8dcae6367559~ffd1977a12e823661da018fa3c662155db9cdb0f (got some necessary fixes) sitting in my local repo for a couple of years, waiting to get out there.  The changed bin/md-to-html5 makes it much easier to generated a properly framed web page from more or less any Markdown file.﻿

levitte commented:
https://beta.openssl.org/ has a run of this PR.  Everything isn't quite right, but close

levitte committed:
Make an exception for the glossary

The glossary is in $(CHECKOUTS)/general-policies/policies/, but is
common for both the general and the technical policies, so we place
it directly in `policies/`.

Since policies/glossary.html is generated from markdown, the breadcrumbs and
sidebar must be duplicated in an added policies/dirdata.yaml.

Fixes #306

levitte committed:
Convert policies/{general,technical}/index.html -> index.md

'cause we can...

paulidale commented:
Aren't the new policies meant to completely replace the old?  The beta doesn't look like that (& to be honest, we're not there yet).

levitte committed:
New script bin/mk-md-titlelist, for enumerating markdown files by title

This works on markdown files only, and will simply grab the first main
title of each file.  It also replaces the `.md` extension with `.html`
in its output.

levitte committed:
Add the necessary files and procedures to generate HTML from policies

The checked out directories $(CHECKOUTS)/technical-policies/policies and
$(CHECKOUTS)/general-policies/policies contain files in markdown format,
which get converted to HTML using pandoc with the new template, and
appropriate dirdata.yaml.

The base infrastructure in policies/general and policies/technical is also
created, and the majority of the links in policies/sidebar.shtml are moved
to policies/general/sidebar.shtml and policies/general/dirdata.yaml.

Fixes #306

levitte committed:
Drop policies/otc-policies.html in favor of policies/technical/

Fixes #306
Fixes openssl/technical-policies#6

levitte committed:
Drop policies/codingstyle.{txt,html} in favor of policies/technical/coding-style.html

Fixes #306

levitte committed:
Drop policies/committers.html in favor of general/committer-policy.html

Fixes #306

levitte committed:
Rework all links that point to General / Technical Policies on github.com

Since all those policies are generated into HTML on our site, we might
as well have all links point there.

levitte commented:
> Aren't the new policies meant to completely replace the old? The beta doesn't look like that (& to be honest, we're not there yet).

We can't *remove* policies without a vote, can we?  So, those that don't have a replacement in general-policies or technical-policies must remain until such a replacement is in place.  I assume that's going to be an incremental thing.

However, to clarify, I've reworked the set of commits, so there are now separate once specifically to drop some of the old files in favor of their replacement.  I suggest doing that same thing in the future, every time a new replacement is added (merged), until we've cleared them all.

levitte committed:
inc/screen.css: don't modify the 'white-space' attribute on 'article a'

This interacts badly with syntax highlighting CSS.

levitte commented:
As seen by the new coding style, we could choose a better syntax hightlighting style.  https://www.garrickadenbuie.com/blog/pandoc-syntax-highlighting-examples/ contains a few examples that pandoc supports, default is said to be pygmets, which isn't suitable for a dark background...

levitte committed:
Generate inc/pandoc-template.html5

We have a inc/pandoc-template.html5, which is a modified pandoc html5
template [^1] from ages ago.  This is problematic, because those
modifications lack more modern pandoc features, such as style values used
for syntax highlighting.

Rather than having our own static template, we can generate it from pandoc's
defaults, in which we inject our own template partials (a term that has
emerged with newer pandoc versions for partial templates that can be used in
a main template).

-----
[^1]: The default pandoc html5 template can be extracted like this:

      ```
      $ pandoc --print-default-template=html5
      ```

levitte committed:
Use the solarized theme for code highlighting with pandoc

This is the theme we use in the blog and that is defined in our screen CSS,
so we ensure that pandoc generates the same theme in its CSS.

The theme definition is derived from this part of the vim solarized theme:
https://github.com/altercation/vim-colors-solarized/blob/528a59f26d12278698bb946f8fb82a63711eec21/colors/solarized.vim#L534-L599

mspncp commented:
> Display the new policies on the site instead of linking to github

I really like this idea! 👍

mspncp commented:
> > Display the new policies on the site instead of linking to github
> 
> I really like this idea! 👍

Would it be a great effort to add an automatically generated „View this page on GitHub“ link at the bottom of the generated page? This would be helpful if people want to suggest changes via an issue or pull request. (I‘ve seen such links e.g. on GitHub Pages.)

paulidale commented on Makefile:
```diff
@@ -204,6 +220,48 @@ docs/mansidebar.html: docs/mansidebar.html.tt Makefile bin/from-tt
+
+######################################################################
+##
+##  Policy page building sexction
```
An 'x' too many.

levitte replied:
Fixed

paulidale commented on Makefile:
```diff
@@ -204,6 +220,48 @@ docs/mansidebar.html: docs/mansidebar.html.tt Makefile bin/from-tt
+		f=$$(basename $$x .md); \
+		cat "$$x" \
+			| sed -E -e 's!https?://github\.com/openssl/(general|technical)-policies/blob/master/policies/(.*)\.md!../\1/\2.html!' \
+			| sed -E -e 's!\.\./general/glossary\.html!../glossary.html!' \
```
Any reason for the second sed instead of sed -e ... -e ...?
Just stylistic?

levitte replied:
Let's say that it's just stylistic...

paulidale commented on policies/general/index.md:
```diff
@@ -0,0 +1,15 @@
+---
+title: OpenSSL General Policies
+---
+These are the general policies and procedures established by the OMC
```
Colon at the end?

levitte replied:
We don't usually do that.  But a period at the very least, yeah

levitte replied:
Done

paulidale commented on policies/technical/index.md:
```diff
@@ -0,0 +1,21 @@
+title: OpenSSL Technical Policies
+subtitle: |
+    First issued 30th September 2020
+    <br />
```
Isn't this just `<br>`?

levitte replied:
I still have XHTML in my fingers :wink:

This is permitted in HTML 5, though...

If you want to enjoy a discussion on the topic, I recommend
https://stackoverflow.com/questions/1946426/html-5-is-it-br-br-or-br

levitte replied:
BTW, not my fault...  it's already there in the file this came from:

https://github.com/openssl/web/blob/68f71e63270307dde5343cfc5b0f31b49cbd3167/policies/otc-policies.html#L15-L16

paulidale replied:
You added a space, so it's obviously your fault :)

paulidale commented on policies/technical/index.md:
```diff
@@ -0,0 +1,21 @@
+    Moved to GitHub Nov 2021
+---
+These are the technical policies and procedures established by the OTC
+in accordance with the [project bylaws]
```
Again a colon at the end to introduce the list properly.

levitte replied:
Or a period...

paulidale reviewed (commented):
Looking decent.

levitte commented:
All review comments addressed so far

paulidale commented on policies/technical/index.md:
```diff
@@ -0,0 +1,21 @@
+They are maintained in the [openssl/technical-policies] repository on GitHub.
+
+
+[project bylaws]: ../omc-bylaws.html
```
Would this be better referencing the glossary?

[project bylaws]: ../glossary.html#bylaws

levitte replied:
....  so any link within policies/ will have to go through the glossary?  Not sure I agree with the forced indirection

levitte replied:
I can, though, if you insist

paulidale replied:
This was the only one that stood out as possibly benefiting from the indirection.  I'm not going to insist however.

The side bar will still have a direct link I believe.

levitte commented:
So, final review?

paulidale reviewed (approved):
Another set of eyes would be good ...

levitte commented:
Ping

levitte commented:
Rebased to fix conflicts

levitte commented:
Ping!

mattcaswell commented on Makefile:
```diff
@@ -216,6 +233,47 @@ docs/fips.inc: $(wildcard docs/fips/*) bin/mk-filelist
+policies/technical/index.html: \
+	policies/technical/index.md policies/technical/index.inc
+
+.PHONY: technical-policies
```
Should be general-policies

levitte replied:
Fixed

t8m commented on Makefile:
```diff
@@ -76,6 +79,12 @@ SIMPLEDOCS = docs/faq.inc docs/fips.inc \
+all_GENERAL_POLICIES=$(wildcard $(CHECKOUTS)/general-policies/policies/*.md)
+all_TECHNICAL_POLICIES=$(wildcard $(CHECKOUTS)/technical-policies/policies/*.md)
+GENERAL_POLICIES=$(filter-out $(CHECKOUTS)/general-policies/policies/README.md $(GLOSSARY),$(all_GENERAL_POLICIES))
+TECHNICAL_POLICIES=$(filter-out $(CHECKOUTS)/general-policies/policies/README.md,$(all_TECHNICAL_POLICIES))
```
This does not make sense to me as your're filtering out README.md from general policies from technical policies which does not contain the README.md at all but if it was there it would have technical-policies in the checkout path.

t-j-h replied:
That should be fixed.

levitte replied:
Fixed by changing `general-policies` to `technical-policies`.  Other than that, I'd prefer to keep this, just in case someone adds a README.md at some point

mattcaswell commented on Makefile:
```diff
@@ -216,6 +233,47 @@ docs/fips.inc: $(wildcard docs/fips/*) bin/mk-filelist
+		d=$$(dirname $$x); \
+		f=$$(basename $$x .md); \
+		cat "$$x" \
+			| sed -E -e 's!https?://github\.com/openssl/(general|technical)-policies/blob/master/policies/(.*)\.md!../\1/\2.html!' \
```
Side note (not for this PR): This makes me wonder whether the original policies should actually be updated to reference the website directly rather than github.

levitte replied:
Nah.  If you look at them on github, you probably want to continue walking around on github

mattcaswell commented on bin/md-to-html5:
```diff
@@ -1,15 +1,180 @@
+the processing result for each of them is written to a corresponding file
+without the '.md' suffix.  All given file names must have the '.md' suffix.
+
+In both usages, one can also set template variables with with the form
```
Typo: with with

levitte replied:
Fixed

mattcaswell commented on bin/md-to-html5:
```diff
@@ -1,15 +1,180 @@
+
+In usage 2, the templates are read from the files given as argument, and
+the processing result for each of them is written to a corresponding file
+without the '.md' suffix.  All given file names must have the '.md' suffix.
```
Perhaps add that -o is not allowed in this usage.

levitte replied:
Fixed

t-j-h commented on policies/technical/index.md:
```diff
@@ -0,0 +1,21 @@
+---
+title: OpenSSL Technical Policies
+subtitle: |
```
There is no point in showing First issued and Moved to GitHub here - this is now just a list - so no need for an attempt at some form of edit history for the *list*.

levitte replied:
I have zero issues dropping that.  Done

t8m reviewed (changes requested):
Just one minor thing found.

t-j-h commented:
Approved with or without fixing up the nit raised (dropping pointless date stuff).

I also agree with @t8m feedback

levitte committed:
fixup! Add the necessary files and procedures to generate HTML from policies

levitte committed:
fixup! Convert policies/{general,technical}/index.html -> index.md

levitte commented:
Review comments addressed

levitte commented:
@t8m, satisfied?

mattcaswell commented on bin/mk-md-titlelist:
```diff
@@ -0,0 +1,40 @@
+        $lastline=$_;
+    }
+    die "Undefined title in $l"
+        unless defined $title;
```
Can this ever happen given we have a default value for `$title`?

levitte replied:
Yes.  If a markdown file starts with a line of equal signs, `$lastline` will be `undef`, and `$title` will therefore become `undef`

levitte committed:
fixup! Rework the markdown to HTML script and template

levitte committed:
fixup! Add the necessary files and procedures to generate HTML from policies

levitte commented:
cfa00986aa96179b40b33e04d04b461132a5358a Rework the markdown to HTML script and template
24a70808948946553f92e62aa6553f2217b474c5 New script bin/mk-md-titlelist, for enumerating markdown files by title
a6adc71de7fe7654e20025c24ccd3dd274d92269 Add the necessary files and procedures to generate HTML from policies
b500ce5c38b6944b0a8ca5250b641c9778053adf Drop policies/otc-policies.html in favor of policies/technical/
b2c7f88e0a1d3aae9cd9e8c1e2e8be501b232caf Drop policies/codingstyle.{txt,html} in favor of policies/technical/coding-style.html
9fcf5661d5c427c42537f28e20c3bf369115ea84 Drop policies/committers.html in favor of general/committer-policy.html
2321b204ef53c5a4503a818aede0401d69acbe72 Make an exception for the glossary
de5c548e809b92e172dfc353ed9e095a04a6e39f Convert policies/{general,technical}/index.html -> index.md
cd366c982506776b15325e039b9891fe3b6f702b Rework all links that point to General / Technical Policies on github.com
d4c7508c91045368714e318f0d0437fe4015499d inc/screen.css: don't modify the 'white-space' attribute on 'article a'
a53269e8550a5a5abacd5d826a7166729f59bd0a Generate inc/pandoc-template.html5
a2c6bd0c1de6c35a7cf92b85d61182cf4242a6c4 Use the solarized theme for code highlighting with pandoc
